website/addons/onedrive/model.py

```diff
@@ -0,0 +1,258 @@
+    default_scopes = ['wl.basic wl.signin onedrive.readwrite wl.offline_access']
+
+    _auth_client = OneDriveAuthClient()
+    _drive_client = OneDriveClient()
+
+    def handle_callback(self, response):        
+        """View called when the Oauth flow is completed. Adds a new OneDriveUserSettings
+        record to the user and saves the user's access token and account info.
+        """
+        
+        userInfoRequest = requests.get(("{}me?access_token={}").format(settings.MSLIVE_API_URL, response['access_token']))
+        
+        logger.debug("userInfoRequest:: %s", repr(userInfoRequest))
+        
+        userInfo = userInfoRequest.json()
+        logger.debug("userInfo:: %s", repr(userInfo))
+
+        return {
+            'provider_id': userInfo['id'],
+            'display_name': userInfo['name'],
+            'profile_url': userInfo['link']
+        }
+        
+    def _refresh_token(self, access_token, refresh_token):
+        """ Handles the actual request to refresh tokens
+
+        :param str access_token: Access token (oauth key) associated with this account
+        :param str refresh_token: Refresh token used to request a new access token
+        :return dict token: New set of tokens
+        """
+        client = self._auth_client
+        if refresh_token:
+            token = client.refresh(access_token, refresh_token)
+            return token
+        else:
+            return False
+        
+    def fetch_access_token(self, force_refresh=False):
+        self.refresh_access_token(force=force_refresh)
+        return self.account.oauth_key
+
+    def refresh_access_token(self, force=False):
+        """ If the token has expired or will soon, handles refreshing and the storage of new tokens
+
+        :param bool force: Indicates whether or not to force the refreshing process, for the purpose of ensuring that authorization has not been unexpectedly removed.
+        """
+        if self._needs_refresh() or force:
+            token = self._refresh_token(self.account.oauth_key, self.account.refresh_token)
+            self.account.oauth_key = token['access_token']
+            self.account.refresh_token = token['refresh_token']
+            self.account.expires_at = datetime.utcfromtimestamp(token['expires_at'])
+            self.account.save()
+
+    def _needs_refresh(self):
+        if self.account.expires_at is None:
+            return False
+        return (self.account.expires_at - datetime.utcnow()).total_seconds() < settings.REFRESH_TIME
+
+class OneDriveUserSettings(AddonOAuthUserSettingsBase):
+    """Stores user-specific onedrive information
+    """
+    oauth_provider = OneDrive
+    serializer = OneDriveSerializer
+#     myBase = AddonOAuthUserSettingsBase
+
+
+class OneDriveNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+
+    oauth_provider = OneDrive
+    serializer = OneDriveSerializer
+
+    foreign_user_settings = fields.ForeignField(
```

Comment:
In other addons this field is a pointer to a legacy database collection and can be removed here.